### PR TITLE
Add a function to primitive assembler that draws vertical arrows

### PR DESCRIPTION
**Pull request by Mahmood-Darwish**
Primitive assembler now has a function that allows drawing vertical arrows, the height, width, head height, and head width ratios are all modifiable.

Bug: http://b/236945179

### CODE REVIEW

File: src/OrbitGl/Track.cpp [hunk 1]
```diff
@@ -12,7 +12,6 @@
 #include "CoreMath.h"
 #include "Geometry.h"
 #include "GlCanvas.h"
-#include "OrbitBase/ThreadUtils.h"
```
Comment (beckerhe) [inline src/OrbitGl/Track.cpp:15]: Nit: This is an unrelated change. We try to keep PRs as focussed as possible which means this ideally shouldn't be in this PR.

**Review by beckerhe — COMMENTED** [review]
This looks good to me.

More of a question to @karupayun: Can this be unit tested (easily)?

File: src/OrbitGl/PrimitiveAssembler.h [hunk 2]
```diff
@@ -98,6 +98,9 @@ class PrimitiveAssembler {
 
   void AddCircle(const Vec2& position, float radius, float z, Color color);
 
+  void AddVerticalArrow(float x, float y_start, float y_end, float z, uint32_t width,
```
Comment (karupayun) [inline src/OrbitGl/PrimitiveAssembler.h:101]: I would be consistent with `AddVerticalLine` here, so using `pos` and `size` (or `lenght` if you just want to include the vertical lenght). But using `pos` and `size` could help you to not need to include `width` and saving a parameter.

File: src/OrbitGl/PrimitiveAssembler.h [hunk 3]
```diff
@@ -98,6 +98,9 @@ class PrimitiveAssembler {
...
+                        float head_length_ratio, float head_width_ratio, Color color);
```
Comment (karupayun) [inline src/OrbitGl/PrimitiveAssembler.h:102]: I'd include these 2 parameter in a simple Vec2 called `arrow_head_size`

Comment (karupayun) [reply src/OrbitGl/PrimitiveAssembler.h:102]: Also, why ratio? Is the head of the arrow be longer depending on the lenght of the arrow? I'd have the same size independent of the lenght.

In addition, I think that giving the absolute values could be more easy to reuse this method.

File: src/OrbitGl/PrimitiveAssembler.cpp [hunk 4]
```diff
@@ -241,6 +244,29 @@ void PrimitiveAssembler::AddCircle(const Vec2& position, float radius, float z,
...
 
+void PrimitiveAssembler::AddVerticalArrow(float x, float start_y, float end_y, float z,
+                                          uint32_t width, float head_length_ratio,
+                                          float head_width_ratio, Color color) {
+  float min_y = std::min(start_y, end_y);
+  float max_y = std::max(start_y, end_y);
+
+  float head_length = std::abs(start_y - end_y) * head_length_ratio;
+  if (max_y == end_y) {
+    head_length = -head_length;
```
Comment (karupayun) [inline src/OrbitGl/PrimitiveAssembler.cpp:255]: Since you are just using `head_length + end_y` 4 times and not directly `head_length`, I will directly create a variable for that

Something like that (including the change I proposed to the .h file):
`float head_arrow_start_y = end_y + arrow_head_size[1] * (max_y == end_y ? -1 : 1);`

File: src/OrbitGl/PrimitiveAssembler.cpp [hunk 5]
```diff
@@ -241,6 +244,29 @@ void PrimitiveAssembler::AddCircle(const Vec2& position, float radius, float z,
...
+  }
+  float head_width = width * head_width_ratio;
+  Triangle arrow_head(Vec2(x, end_y), Vec2(x - head_width, end_y + head_length),
+                      Vec2(x + head_width, end_y + head_length));
+  AddTriangle(arrow_head, z, color);
+
+  min_y = std::min(start_y, end_y + head_length);
```
Comment (karupayun) [inline src/OrbitGl/PrimitiveAssembler.cpp:262]: maybe `arrow_body_min_y`

File: src/OrbitGl/PrimitiveAssembler.cpp [hunk 6]
```diff
@@ -241,6 +244,29 @@ void PrimitiveAssembler::AddCircle(const Vec2& position, float radius, float z,
...
+  max_y = std::max(start_y, end_y + head_length);
```
Comment (karupayun) [inline src/OrbitGl/PrimitiveAssembler.cpp:263]: and `arrow_body_max_y`

**Review by karupayun — COMMENTED** [review]
Nice! Happy to see step by step how this arrow is being created :)

Looking forward to have a screenshot of it!

I left a few comments

File: src/OrbitGl/Track.cpp (hunk 1 above)
Comment (karupayun) [reply src/OrbitGl/Track.cpp:15]: > This looks good to me.
> 
> More of a question to @karupayun: Can this be unit tested (easily)?

This function is a perfect candidate for `TEST(PrimitiveAssembler, ComplexShapes)` in `PrimitiveAssemblerTest.cpp`.

**Comment by karupayun** [comment]
> This looks good to me.
> 
> More of a question to @karupayun: Can this be unit tested (easily)?

This function is a perfect candidate for `TEST(PrimitiveAssembler, ComplexShapes)` in `PrimitiveAssemblerTest.cpp`.

File: src/OrbitGl/PrimitiveAssembler.h (hunk 2 above)
Comment (Mahmood-Darwish) [reply src/OrbitGl/PrimitiveAssembler.h:101]: I get this but the ending position could be higher or lower than the starting position. I would only get xs and ys for the thread slices and I would like to use them as is, instead of having to do some calculations before calling the function. 

I think I'll just overload the function with a signature similar to vertical line instead of changing this one.

File: src/OrbitGl/Track.cpp (hunk 1 above)
Comment (Mahmood-Darwish) [reply src/OrbitGl/Track.cpp:15]: Understood. I'll leave it this time, but I'll make sure to separate PRs next time.

File: src/ClientData/DataManager.cpp [hunk 7]
```diff
@@ -57,7 +57,7 @@ void DataManager::set_selected_thread_id(uint32_t thread_id) {
 void DataManager::set_selected_thread_state_slice(
     std::optional<orbit_client_data::ThreadStateSliceInfo> selected_thread_state_slice) {
   ORBIT_CHECK(std::this_thread::get_id() == main_thread_id_);
-  selected_thread_state_slice_ = std::move(selected_thread_state_slice);
+  selected_thread_state_slice_ = selected_thread_state_slice;
```
[the remark this answers is not in the document]
Comment (Mahmood-Darwish) [reply src/ClientData/DataManager.cpp:60]: Understood. I'll leave it this time, but I'll make sure to separate PRs next time.

File: src/OrbitGl/PrimitiveAssembler.h (hunk 2 above)
Comment (karupayun) [reply src/OrbitGl/PrimitiveAssembler.h:101]: I think it's better to do the calculations in your method and use just a standardized method here. I don't have a strong opinion between this and the VerticalLine but yes about being consistent and using the same pattern in both places.

File: src/OrbitGl/PrimitiveAssemblerTest.cpp [hunk 8]
```diff
@@ -18,6 +18,8 @@ const Vec2 kTopRight{5, 0};
 const Vec2 kBottomRight{5, 5};
 const Vec2 kBottomLeft{0, 5};
 const Vec2 kBoxSize{5, 5};
+const Vec2 kArrowSize{2, 4};
+const Vec2 kArrowHeadSize{3, 2};
```
Comment (karupayun) [inline src/OrbitGl/PrimitiveAssemblerTest.cpp:22]: It's a bit strange that the `arrow_head` is bigger than the `total_arrow_size`. As I told you above, consider instead having `body_size` and `head_size`

**Review by karupayun — COMMENTED** [review]
Hey, I left some comments. A few of them are repeated points. :)

File: src/OrbitGl/PrimitiveAssemblerTest.cpp [hunk 9]
```diff
@@ -154,28 +156,31 @@ TEST(PrimitiveAssembler, ComplexShapes) {
...
   primitive_assembler_tester.StartNewFrame();
 
-  // Draw arrow pointing up
-  primitive_assembler_tester.AddVerticalArrow(kBottomRight, kArrowSize, 0, kFakeColor,
-                                              PrimitiveAssembler::ArrowDirection::kUp,
-                                              kArrowHeadSize);
+  // AddVerticalArrow (arrow pointing up) -> 1 box and 1 Triangle
+  primitive_assembler_tester.AddVerticalArrow(kArrowStartingPos, kArrowBodySize, kArrowHeadSize, 0,
+                                              kFakeColor, PrimitiveAssembler::ArrowDirection::kUp);
+  float arrow_width2 = std::max(kArrowHeadSize[0], kArrowBodySize[0]);
```
Comment (karupayun) [inline src/OrbitGl/PrimitiveAssemblerTest.cpp:176]: Nit: I didn't need/consider that before but you can just use arrow_width in both places by scoping each part.

```
{
   // AddVerticalArrow (arrow pointing down) -> 1 box and 1 Triangle
   float arrow_width = ....
}

{
   // AddVerticalArrow (arrow pointing up) -> 1 box and 1 Triangle
   float arrow_width = ....
}

```

**Review by karupayun — APPROVED** [review]
LGTM! Thanks for addressing all my comments :)

**Review by beckerhe — APPROVED** [review]
Awesome. Great. LGTM!!!
